From 65dd753c24629938fda3f6673c1cde898d9f7248 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Thu, 14 Apr 2016 14:12:31 +0300
Subject: [PATCH 1046/1240] fix: clock: a3700: Remove clock presets from DT

- Remove clock presets from DT and move to defconfig
  This change is required since the clock preset has moved
  to TIM and cannot be changed on the fly, but only
  on compilation time.
- Fixes SYSTEMSW-2433

Change-Id: I2b0043cae00ef496402750d9a2f3fc60f611a647
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/29046
Reviewed-by: Haim Boot <hayim@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
---
 Makefile                                       | 12 ++++++++++--
 arch/arm/cpu/armv8/armada3700/Kconfig          | 26 +++++++++++++++++++++++++
 arch/arm/cpu/armv8/armada3700/mvebu_clock.c    | 24 ++++++++++-------------
 arch/arm/dts/armada-3700-customer0.dts         |  4 ----
 arch/arm/dts/armada-3700-db.dts                |  4 ----
 arch/arm/dts/armada-3700.dtsi                  |  6 ------
 arch/arm/include/asm/arch-armada3700/clock.h   |  8 +++++---
 doc/device-tree-bindings/clock/mvebu-clock.txt | 26 -------------------------
 include/dt-bindings/clock/freq_data.h          | 27 --------------------------
 include/fdtdec.h                               |  1 -
 lib/fdtdec.c                                   |  1 -
 11 files changed, 51 insertions(+), 88 deletions(-)
 delete mode 100644 include/dt-bindings/clock/freq_data.h

diff --git a/Makefile b/Makefile
index 6e8475a..fdad187 100644
--- a/Makefile
+++ b/Makefile
@@ -852,8 +852,16 @@ TIM2PHEX	:= $(srctree)/scripts/tim2phex.pl
 TIM2IMG		:= $(srctree)/scripts/tim2img.pl
 TIMBUILD	:= $(srctree)/scripts/buildtim.sh
 CLOCKSPATH	:= $(srctree)/tools/wtp
-CLOCKSPRESET	:= $(shell find $(srctree) -name $(CONFIG_DEFAULT_DEVICE_TREE).dts | \
-			xargs grep preset | sed -n 's/.*<\([^ ]*\)>;/\1/p')
+
+ifeq ($(CONFIG_PRESET_CPU_600_DDR_600),y)
+CLOCKSPRESET	:= PRESET_CPU_600_DDR_600
+else ifeq ($(CONFIG_PRESET_CPU_800_DDR_800),y)
+CLOCKSPRESET	:= PRESET_CPU_800_DDR_800
+else ifeq ($(CONFIG_PRESET_CPU_1000_DDR_800),y)
+CLOCKSPRESET	:= PRESET_CPU_1000_DDR_800
+else
+CLOCKSPRESET	:= PRESET_INVALID
+endif
 
 ifeq ($(CONFIG_MVEBU_SPI_BOOT),y)
 BOOTDEV		:= SPINOR
diff --git a/arch/arm/cpu/armv8/armada3700/Kconfig b/arch/arm/cpu/armv8/armada3700/Kconfig
index d4fc13d..64a3d91 100644
--- a/arch/arm/cpu/armv8/armada3700/Kconfig
+++ b/arch/arm/cpu/armv8/armada3700/Kconfig
@@ -50,3 +50,29 @@ config OSLO_START_ADDR
 
 endif
 
+choice
+	prompt "CPU and DDR Clock preset"
+	default PRESET_CPU_800_DDR_800
+
+config PRESET_CPU_600_DDR_600
+	bool "CPU at 600 MHz, DDR at 600 MHz"
+	help
+	  Run system with the following clock
+	  tree configuration preset:
+	  CPU at 600 MHz, DDR3 at 600 MHz
+
+config PRESET_CPU_800_DDR_800
+	bool "CPU at 800 MHz, DDR at 800 MHz"
+	help
+	  Run system with the following clock
+	  tree configuration preset:
+	  CPU at 800 MHz, DDR3 at 800 MHz
+
+config PRESET_CPU_1000_DDR800
+	bool "CPU at 1000 MHz, DDR at 800 MHz"
+	help
+	  Run system with the following clock
+	  tree configuration preset:
+	  CPU at 1000 MHz, DDR3 at 800 MHz
+
+endchoice
diff --git a/arch/arm/cpu/armv8/armada3700/mvebu_clock.c b/arch/arm/cpu/armv8/armada3700/mvebu_clock.c
index db0ca74..f0c72dc 100644
--- a/arch/arm/cpu/armv8/armada3700/mvebu_clock.c
+++ b/arch/arm/cpu/armv8/armada3700/mvebu_clock.c
@@ -472,24 +472,20 @@ u32 set_clocks(u32 cpu_clk_mhz, u32 ddr_clk_mhz, u32 tbg_a_kvco_mhz, u32 tbg_b_k
 
 int init_clock(void)
 {
-	int node, count, idx, ret;
-	const void *blob = gd->fdt_blob;
-	int tbl_sz = sizeof(a3700_clock_configs)/sizeof(a3700_clock_configs[0]);
+	int idx, ret;
 	u32 vdd_val;
 
 	debug_enter();
 
-	count = fdtdec_find_aliases_for_id(blob, "freq", COMPAT_MVEBU_A3700_FREQ, &node, 1);
-	if (count == 0) {
-		error("The frequency preset is not defined in DT, using default\n");
-		idx = MVEBU_A3700_DEF_CLOCK_PRESET_IDX;
-	} else {
-		idx = fdtdec_get_int(blob, node, "preset", MVEBU_A3700_DEF_CLOCK_PRESET_IDX);
-		if ((idx >= tbl_sz) || (idx < 0)) {
-			error("Unsupported frequency preset in DT (%d), using default\n", idx);
-			idx = MVEBU_A3700_DEF_CLOCK_PRESET_IDX;
-		}
-	}
+#if defined(CONFIG_PRESET_CPU_600_DDR_600)
+	idx = MVEBU_A3700_PRESET_IDX_CPU_600_DDR_600;
+#elif defined(CONFIG_PRESET_CPU_800_DDR_800)
+	idx = MVEBU_A3700_PRESET_IDX_CPU_800_DDR_800;
+#elif defined(CONFIG_PRESET_CPU_1000_DDR_800)
+	idx = MVEBU_A3700_PRESET_IDX_CPU_1000_DDR_800;
+#else
+#error "Bad clock preset!"
+#endif
 
 	printf("Setting clocks to CPU=%dMHz and DDR=%dMHz\n",
 		a3700_clock_configs[idx].cpu_freq_mhz, a3700_clock_configs[idx].ddr_freq_mhz);
diff --git a/arch/arm/dts/armada-3700-customer0.dts b/arch/arm/dts/armada-3700-customer0.dts
index c69dd98..6d423fb 100644
--- a/arch/arm/dts/armada-3700-customer0.dts
+++ b/arch/arm/dts/armada-3700-customer0.dts
@@ -9,10 +9,6 @@
 	compatible = "marvell,armada-3700-db0", "marvell,armada-3700";
 
 	soc {
-		freq {
-			preset = <PRESET_CPU_600_DDR_600>;
-		};
-
 		internal-regs {
 			ddr-mac {
 				cs_count = <1>;
diff --git a/arch/arm/dts/armada-3700-db.dts b/arch/arm/dts/armada-3700-db.dts
index d706d55..87f0668 100644
--- a/arch/arm/dts/armada-3700-db.dts
+++ b/arch/arm/dts/armada-3700-db.dts
@@ -29,10 +29,6 @@
 	board_id = <A3700_DB_ID>;
 
 	soc {
-		freq {
-			preset = <PRESET_CPU_800_DDR_800>;
-		};
-
 		internal-regs {
 			ddr-mac {
 				cs_count = <1>;
diff --git a/arch/arm/dts/armada-3700.dtsi b/arch/arm/dts/armada-3700.dtsi
index 05a3bc5..761a97c 100644
--- a/arch/arm/dts/armada-3700.dtsi
+++ b/arch/arm/dts/armada-3700.dtsi
@@ -2,7 +2,6 @@
 #define IO_ATTR(max_win, max_remap, remap_size, win_offset) (((max_win) << 24) | ((max_remap) << 16) | ((remap_size) << 8) | (win_offset))
 
 #include <dt-bindings/comphy/comphy_data.h>
-#include <dt-bindings/clock/freq_data.h>
 / {
 	model = "Marvell Armada 3700 Development Board";
 	compatible = "marvell,armada-3700";
@@ -15,11 +14,6 @@
 	#address-cells = <1>;
 	#size-cells = <0>;
 
-		freq {
-			compatible = "marvell,a3700-freq";
-			preset = <PRESET_CPU_600_DDR_600>;
-		};
-
 		internal-regs {
 			compatible = "marvell,internal-regs";
 			#address-cells = <1>;
diff --git a/arch/arm/include/asm/arch-armada3700/clock.h b/arch/arm/include/asm/arch-armada3700/clock.h
index 8abf1f3..aa92917 100644
--- a/arch/arm/include/asm/arch-armada3700/clock.h
+++ b/arch/arm/include/asm/arch-armada3700/clock.h
@@ -275,9 +275,11 @@ struct a3700_clock_cfg {
 	struct a3700_sb_clock_cfg	sb_clk_cfg;
 };
 
-/* This index points to configuration selected in MVEBU_A3700_CLOCK_CFGS
-   when DT entry is missing or invalid */
-#define MVEBU_A3700_DEF_CLOCK_PRESET_IDX	1
+/* Indexes to MVEBU_A3700_CLOCK_CFGS array */
+#define MVEBU_A3700_PRESET_IDX_CPU_600_DDR_600		0
+#define MVEBU_A3700_PRESET_IDX_CPU_400_DDR_600		1
+#define MVEBU_A3700_PRESET_IDX_CPU_1000_DDR_800		2
+#define MVEBU_A3700_PRESET_IDX_CPU_800_DDR_800		3
 
 /* Init values for the static clock configurations array */
 /*
diff --git a/doc/device-tree-bindings/clock/mvebu-clock.txt b/doc/device-tree-bindings/clock/mvebu-clock.txt
index fa1a899..da171c0 100644
--- a/doc/device-tree-bindings/clock/mvebu-clock.txt
+++ b/doc/device-tree-bindings/clock/mvebu-clock.txt
@@ -8,29 +8,3 @@ SoC specific:
 	- compatible: Should be "marvell,tclk".
 	- reg = <0x18600 0x4>;
 		Base address and size of sample at reset register.
-
-For the Armada 3700 platform, the DT allows selecting one of the
-CPU/DDR frequency presets under the SOC node.
-
-freq: Define the clock preset for CPU/DDR
-	compatible = "marvell,a3700-freq";
-	preset = <freq_preset>;
-		Supported presets:
-		PRESET_CPU_600_DDR_600  - CPU 600MHz,  DDR 600MHz
-		PRESET_CPU_400_DDR_600  - CPU 400MHz,  DDR 600MHz
-		PRESET_CPU_1000_DDR_800 - CPU 1000MHz, DDR 800MHz
-		PRESET_CPU_800_DDR_800  - CPU 800MHz,  DDR 800MHz
-
-Armada 3700 example:
-soc {
-	freq {
-		compatible = "marvell,a3700-freq";
-		preset = <PRESET_CPU_400_DDR_600>;
-	};
-	....
-	....
-	internal-regs {
-		....
-		....
-	};
-};
diff --git a/include/dt-bindings/clock/freq_data.h b/include/dt-bindings/clock/freq_data.h
deleted file mode 100644
index 2971f66..0000000
--- a/include/dt-bindings/clock/freq_data.h
+++ /dev/null
@@ -1,27 +0,0 @@
-/*
- * ***************************************************************************
- * Copyright (C) 2015 Marvell International Ltd.
- * ***************************************************************************
- * This program is free software: you can redistribute it and/or modify it
- * under the terms of the GNU General Public License as published by the Free
- * Software Foundation, either version 2 of the License, or any later version.
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program.  If not, see <http://www.gnu.org/licenses/>.
- * ***************************************************************************
- */
-
-#ifndef _FREQ_DATA_H_
-#define _FREQ_DATA_H_
-
-#define PRESET_CPU_600_DDR_600		0
-#define PRESET_CPU_400_DDR_600		1
-#define PRESET_CPU_1000_DDR_800		2
-#define PRESET_CPU_800_DDR_800		3
-
-#endif /* _FREQ_DATA_H_ */
diff --git a/include/fdtdec.h b/include/fdtdec.h
index 2706a19..ef217cd 100644
--- a/include/fdtdec.h
+++ b/include/fdtdec.h
@@ -157,7 +157,6 @@ enum fdt_compat_id {
 	COMPAT_MVEBU_DDR_PHY,
 	COMPAT_MVEBU_ADVK_PCIE,
 	COMPAT_MVEBU_XENON_MMC,
-	COMPAT_MVEBU_A3700_FREQ,
 	COMPAT_MVEBU_MDIO,
 	COMPAT_MVEBU_PXA3XX_NAND,
 	COMPAT_MVEBU_UTMI_PHY,
diff --git a/lib/fdtdec.c b/lib/fdtdec.c
index 722bca5..cf6acfd 100644
--- a/lib/fdtdec.c
+++ b/lib/fdtdec.c
@@ -119,7 +119,6 @@ static const char * const compat_names[COMPAT_COUNT] = {
 	COMPAT(MVEBU_DDR_PHY, "marvell,mvebu-ddr-phy"),
 	COMPAT(MVEBU_ADVK_PCIE, "marvell,advk-pcie"),
 	COMPAT(MVEBU_XENON_MMC, "marvell,xenon-sdhci"),
-	COMPAT(COMPAT_MVEBU_A3700_FREQ, "marvell,a3700-freq"),
 	COMPAT(MVEBU_MDIO, "marvell,mvebu-mdio"),
 	COMPAT(MVEBU_PXA3XX_NAND, "marvell,mvebu-pxa3xx-nand"),
 	COMPAT(MVEBU_UTMI_PHY, "marvell,mvebu-utmi-2.6.0"),
-- 
1.9.1

